sympy/matrices/random.py

```diff
@@ -0,0 +1,2343 @@
+        elementary matrices to build the invertible matrix
+    units : tuple or list of symbols (optional)
+        default values for random choosen scalar of
+        elementary matrices to build the invertible matrix
+        and non-zero diagonal entries
+        which are assumed to have a multiplicative inverse
+    length : integer (optional with default 2 * **dim**)
+        number of invertible matrices to build the resulting matrix.
+    See Also
+    ========
+    elementary
+    square
+
+    """
+
+    if length == 0:
+        return regular_to_singular(eye(dim), rank)
+    scalars = scalars or (0,)
+    units = units or (1,)
+    length = length or 2 * dim
+    row_indicies = [_sample(range(dim)) for _ in range(length)]
+    indicies = [(i, _sample(range(i, dim))) for i in row_indicies]
+    scalars = [_sample(units) if i == j
+               else _sample(scalars) for i, j in indicies]
+    items = [elementary(dim, ix, s) for ix, s in zip(indicies, scalars)]
+    return regular_to_singular(Mul(*items), rank)
+
+
+def square(dim,
+           rank=None,
+           scalars=_elementary_scalars,
+           units=_elementary_units,
+           length=None):
+    r"""a square matrix n x n with a given rank
+
+    Explanation
+    ===========
+    A full rank square matrix has **rank=dim**.
+    Such a matrix is constructed as
+    product of :func:`elementary` matrices.
+
+    If **rank** is less than **dim**,
+    it is constructed from a full rank matrix $\mathbf{S}$
+    by choosing randomly $r$ basis columns
+    and build $n-r$ columes as a linear combinations
+    of the basis colums.
+
+    Such a square matrix $\mathbf{S}$ represents an endomorphism
+    $$f_S:V \rightarrow V, v \mapsto \mathbf{S}\cdot v$$
+    of a $n$ dimensional vector space $V$.
+
+    Parameters
+    ==========
+    dim : integer
+        dimension of matrix
+    rank : integer (optional with default **dim**)
+        rank of matrix
+    scalars : tuple or list of symbols (optional)
+        default values for random choosen scalar of
+        elementary matrices to build the matrix
+    units : tuple or list of symbols (optional)
+        default values for random choosen scalar of
+        elementary matrices to build the matrix
+        which are assumed to have a multiplicative inverse
+    length : integer (optional with default 2 * **dim**)
+        number of invertible matrices
+        to build the resulting matrix.
+
+    Examples
+    ========
+    >>> from sympy.core.random import seed
+    >>> from sympy.matrices.random import square
+    >>> seed(1)
+
+    >>> square(3)
+    Matrix([
+    [-1, -1, -1],
+    [-2, -1, -1],
+    [ 0,  0,  1]])
+
+    See Also
+    ========
+    triangular
+    invertible
+    singular
+
+    """
+
+    if length == 0:
+        return regular_to_singular(eye(dim), rank)
+
+    length = length or 2 * dim
+    lwr = triangular(dim, None, scalars, units,
+                     int(length / 2))
```

Comment:
`int(length / 2)` is `length // 2` (which avoids the use of floats as an intermediary).

Reply:
Nice.